workflow-ui/core-common/src/main/java/com/squareup/workflow1/ui/Container.kt

```diff
@@ -0,0 +1,72 @@
+package com.squareup.workflow1.ui
+
+import com.squareup.workflow1.ui.Compatible.Companion.keyFor
+
+/**
+ * A rendering type comprised of a set of other renderings.
+ *
+ * Why two parameter types? The separate [BaseT] type allows implementations
+ * and sub-interfaces to constrain the types that [map] is allowed to
+ * transform [ContentT] to. E.g., it allows `FooWrapper<S: Screen>` to declare
+ * that [map] is only able to transform `S` to other types of `Screen`.
+ *
+ * @param BaseT the invariant base type of the contents of such a container,
+ * usually [Screen] or [Overlay][com.squareup.workflow1.ui.container.Overlay].
+ * It is common for the [Container] itself to implement [BaseT], but that is
+ * not a requirement. E.g., [ScreenOverlay][com.squareup.workflow1.ui.container.ScreenOverlay]
+ * is an [Overlay][com.squareup.workflow1.ui.container.Overlay], but it
+ * wraps a [Screen].
+ *
+ * @param ContentT the specific subtype of [BaseT] collected by this [Container].
+ */
+@WorkflowUiExperimentalApi
+public interface Container<BaseT, out ContentT : BaseT> {
+  public fun asSequence(): Sequence<ContentT>
+
+  /**
+   * Returns a [Container] with the [transform]ed contents of the receiver.
+   * It is expected that an implementation will take advantage of covariance
+   * to declare its own type as the return type, rather than plain old [Container].
+   * This requirement is not enforced because recursive generics are a fussy nuisance.
+   *
+   * For example, suppose we want to create `LoggingScreen`, one that wraps any
+   * other screen to add some logging calls. Its implementation of this method
+   * would be expected to have a return type of `LoggingScreen` rather than `Container`:
+   *
+   *    override fun <D : Screen> map(transform: (C) -> D): LoggingScreen<D> =
+   *      LoggingScreen(transform(content))
+   *
+   * By requiring all [Container] types to implement [map], we ensure that their
+   * contents can be repackaged in interesting ways, e.g.:
+   *
+   *    val childBackStackScreen = renderChild(childWorkflow) { ... }
+   *    val loggingBackStackScreen = childBackStackScreen.map { LoggingScreen(it) }
+   */
+  public fun <ContentU : BaseT> map(transform: (ContentT) -> ContentU): Container<BaseT, ContentU>
+}
+
+/**
+ * A singleton [Container] -- that is, a rendering which wraps a single rendering of type
```

Comment:
nit: "singleton" Container still throws me off as I think, oh - it makes it so that this Rendering is a Singleton; which, i get, is absurd but I need the explanation sentence anyway so why not change what we are summarizing this as?

"A frame [Container]"?

Reply:
```
 * A [Container] rendering that wraps exactly one other rendering, its [content]. These are
 * typically used to "add value" to the [content], e.g. an
 * [EnvironmentScreen][com.squareup.workflow1.ui.container.EnvironmentScreen] that allows
 * changes to be made to the the [ViewEnvironment].
```

Reply:
"the the" :/